README
see COPYING for the license (LGPL)

Dependencies
============
- Reasonably compliant C++ compiler (gcc 3.x, vc 7.1, etc)
- Boost library: See www.boost.org for details and downloads
- CppUnit (optional for test suite): See cppunit.sf.net
- SConsAddons - Addon library for the build system.  https://realityforge.vrsource.org/SconsAddons

Building on Windows
===================
Open the file vc7\cppdom.sln using Visual Studio (version 2002 or newer
is required).  Build the desired targets (usually "Debug - DLL" and
"Release - DLL").  The compiled files will be in the 'lib' subdirectory of
the source tree.

Installation must be performed by hand.  The following commands show a
possible way to install everything correctly:

   mkdir C:\cppdom\include\cppdom
   mkdir C:\cppdom\include\cppdom\ext
   mkdir C:\cppdom\lib
   copy cppdom\*.h C:\cppdom\include\cppdom
   copy cppdom\ext\*.h C:\cppdom\include\cppdom\ext
   copy lib\*.lib C:\cppdom\lib
   copy lib\*.dll C:\cppdom\lib

NOTE: Compiling the OPTIONAL Spirit-based parser (see cppdom\SpiritParser.cpp)
is not currently supported by the Visual Studio solution vc7\cppdom.sln.

Building
========
CppDom uses the scons build system.  You can download scons from www.scons.org.

We also use the SconsAddons library (see above).  You will need its source 
directory in your PYTHONPATH to build.

ex: https://realityforge.vrsource.org/svn/scons-addons/trunk/scons-addons

PYTHONPATH=/home/username/scons-addons/src

NOTE: In the future we hope to use third party branches or svn:externals to 
remove the need to download this separately.

Once you have both of these, the follow steps should be all you need.

Please see 'scons --help' for a list of options for specifying where the
dependencies are located.

type 'scons' to build cppdom shared lib.
type 'scons install' to install on the system
type 'scons install prefix=$HOME/software' to install in your account


ChangeLog
DATE        AUTHOR      CHANGE
---------- -----------  -------------------------------------------------------
2012-07-04 patrickh     Add CMake build.
                        Developed by Ryan Pavlik.

[Version 1.2.0 released - 5.1.2011]============================================

2011-04-23 patrickh     SCons 2.0 is now the minimum required version.
                        Submitted by Doug McCorkle.
2010-04-14 patrickh     Preserve newlines in CDATA.
                        Submitted by: Carsten Neumann
                        VERSION: 1.1.2
2010-03-06 patrickh     Fixed conflicts with the operator<< overload for
                        std::ostream and implicit constructio of
                        cppdom::Attribute objects.
                        Submitted by: Carsten Neumann
                        VERSION: 1.1.1

[Version 1.0.0 released - 3.2.2009]============================================

2007-08-01 patrickh     Updated for SConsAddons changes. On Windows, the
                        build formerly called "hybrid" is now the "debug"
                        build. The DLL is installed in lib\debug. The build
                        formerly called "debug" is now the "debugrt" build (to
                        indicate that it is linked against the Visual C++ debug
                        runtime). The library is still named with "_d".
                        VERSION: 0.7.10
2007-07-25 patrickh     Fixed empty comment processing bug.
2007-07-06 patrickh     Fixed build issues on Mac OS X.
                        VERSION: 0.7.9
2007-06-26 patrickh     Made many build fixes and improvements on Windows.
                        VERSION: 0.7.8
2006-11-07 patrickh     Do not put headers into a versioned directory on
                        Windows.
                        VERSION: 0.7.7
2006-07-13 patrickh     Implemented automatic linking on Windows.
                        VERSION: 0.7.6
2006-07-13 patrickh     Moved the declaration of cppdom::getVersion() to
                        cppdom.h.
2006-07-11 allenb       Expose OptionRepository and cppdom methods in dll.
2006-07-05 allenb       More refactoring of build system.
                        It is not possible to specify all variants to build
                        and build them simultaneously.
2006-06-30 allenb       Refatored the build system to support building on Win32
                        using SCons.  This requires an in-development version
                        of scons-addons to get the EnvironmentBuilder.
2006-06-24 patrickh     Support building universal binaries on Mac OS X.
2006-06-13 allenb       Fixed bug in how NodePtr and DocumentPtr types were
                        defined.  Now explicitly declare Node and Document.
                        VERSION: 0.7.5
2006-06-06 dshipton     Version out includes and library
                        VERSION: 0.7.4
2006-06-05 allenb       Refactored the build system to remove AutoDist.
2006-06-02 dshipton     Install .pc file for pkg-config use
                        VERSION: 0.7.3
2006-02-07 allenb       Refactored the Attributes class to just be an std::map
                        of strings to attributes.
                        Added an attrib() method to Node that returns this map.
                        VERSION: 0.7.2
2006-05-06 allenb       Added ability to set options in OptionRepository.
                        Added saveOptionFile to OptionRepository.
                        Refactored OptionRepository to hold onto Doc node.
                        cppdom::Error now derived from std::exception
                        VERSION: 0.7.1
                        
[Version 0.6.0 released - 6.1.2005]============================================
2005-02-06 allenb       Added hasOption() method to OptionRepository
                        VERSION: 0.5.6
2005-02-04 allenb       Added ability to get children by path.
                        Added extension to assist with using xml files
                        as configuration files.  See: ext/OptionRepository.h
                        VERSION: 0.5.5
2005-02-03 allenb       Added tests for Node interface.
                        Added Node::hasChild() method.

cppdom/config.h
/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil c-basic-offset: 3 -*- */
// vim:cindent:ts=3:sw=3:et:tw=80:sta:
/*************************************************************** cppdom-cpr beg
 * 
 * cppdom was forked from the original xmlpp version 0.6 under the LGPL. This
 * new, branched xmlpp is under the same LGPL (of course) and is being
 * maintained by:
 *      Kevin Meinert   <subatomic@users.sourceforge.net>
 *      Allen Bierbaum  <allenb@users.sourceforge.net>
 *      Ben Scott       <nonchocoboy@users.sourceforge.net>
 *
 * -----------------------------------------------------------------
 *
 * xmlpp - an xml parser and validator written in C++
 * copyright (c) 2000-2001 Michael Fink
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * -----------------------------------------------------------------
 * File:          $RCSfile$
 * Date modified: $Date$
 * Version:       $Revision$
 * -----------------------------------------------------------------
 *
 ************************************************************ cppdom-cpr-end */
/*! \file config.h

  system dependent stuff

*/

#ifndef CPPDOM_CONFIG_H
#define CPPDOM_CONFIG_H

// -----------------------------------
// win32 configuration
#if defined(WIN32) || defined(WIN64)

// switch some warnings off
#  pragma warning( disable: 4786 4275 4251 )

// export/import #define's for building a win32 dll
#  ifdef CPPDOM_EXPORTS
#     define CPPDOM_EXPORT(ret)  __declspec(dllexport) ret __stdcall
#     define CPPDOM_CLASS        __declspec(dllexport)
#  else
#     define CPPDOM_EXPORT(ret)  __declspec(dllimport) ret __stdcall
#     define CPPDOM_CLASS        __declspec(dllimport)

// Use automatic linking when building with Visual C++ and when requested to
// do so. Define either CPPDOM_AUTO_LINK or CPPDOM_DYN_LINK to enable automatic
// linking. The latter specifies that the dynamic library is the one to link
// against automatically. If CPPDOM_DYN_LINK is not defined when using
// automatic linking, then the static library is used.
#     if defined(_MSC_VER) && \
            (defined(CPPDOM_AUTO_LINK) || defined(CPPDOM_DYN_LINK))
#        if defined(_DEBUG)
#           if defined(CPPDOM_DYN_LINK)
#              pragma comment(lib, "cppdom_d.lib")
#           else
#              pragma comment(lib, "cppdom_d_s.lib")
#           endif
#        else  /* ! defined(_DEBUG) */
#           if defined(CPPDOM_DYN_LINK)
#              pragma comment(lib, "cppdom.lib")
#           else
#              pragma comment(lib, "cppdom_s.lib")
#           endif
#        endif /* defined(_DEBUG) */
#     endif /* defined(_MSC_VER) && (defined(CPPDOM_AUTO_LINK) || defined(CPPDOM_DYN_LINK)) */
#  endif /* defined(CPPDOM_EXPORTS) */

#else
#  define CPPDOM_EXPORT(ret) ret
#  define CPPDOM_CLASS
#endif

#ifdef _DEBUG
#define CPPDOM_DEBUG
#endif

// -----------------------------------
#endif


cppdom/cppdom/cppdom.h
/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil c-basic-offset: 3 -*- */
// vim:cindent:ts=3:sw=3:et:tw=80:sta:
/*************************************************************** cppdom-cpr beg
 *
 * cppdom was forked from the original xmlpp version 0.6 under the LGPL. This
 * new, branched xmlpp is under the same LGPL (of course) and is being
 * maintained by:
 *      Kevin Meinert   <subatomic@users.sourceforge.net>
 *      Allen Bierbaum  <allenb@users.sourceforge.net>
 *      Ben Scott       <nonchocoboy@users.sourceforge.net>
 *
 * -----------------------------------------------------------------
 *
 * xmlpp - an xml parser and validator written in C++
 * copyright (c) 2000-2001 Michael Fink
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * -----------------------------------------------------------------
 * File:          $RCSfile$
 * Date modified: $Date$
 * Version:       $Revision$
 * -----------------------------------------------------------------
 *
 ************************************************************ cppdom-cpr-end */
/**
 * @file cppdom.h
 *
 * The main declaration header for all of cppdom.
 */

// prevent multiple includes
#ifndef CPPDOM_CPPDOM_H
#define CPPDOM_CPPDOM_H

#ifdef _MSC_VER
   // disable 'identifier was truncated to 255 characters in debug information' warning
#  pragma warning(disable: 4786)

   // Visual Studio 7 has support for member templates.
#  if _MSC_VER < 1300
#     define CPPDOM_NO_MEMBER_TEMPLATES
#  endif
#endif

// needed includes
#include <string>
#include <sstream>
#include <vector>


// ---- TYPE DEFS for CPPDOM --- //
/** handle to a tagname string in a tagname map */
namespace cppdom
{
   typedef int TagNameHandle;
}

#define CPPDOM_USE_HASH_MAP 1

// Use fastest map available
#if defined(CPPDOM_USE_HASH_MAP) && defined(__GNUC__) && (__GNUC__ >= 3)

#include <ext/hash_map>
#include <map>

namespace std
{ using namespace __gnu_cxx; }

namespace cppdom
{
   struct HashString
   {
      bool operator()(std::string const& str) const
      { return __gnu_cxx::hash<char const *>()(str.c_str()); }
   };

   typedef std::hash_map<TagNameHandle,std::string>   TagNameMap_t;
   typedef std::hash_map<std::string, TagNameHandle, HashString>  NameToTagMap_t;
}

#else

#include <map>
namespace cppdom
{
   typedef std::map<TagNameHandle,std::string>   TagNameMap_t;
   typedef std::map<std::string, TagNameHandle>  NameToTagMap_t;
}
#endif

#include "config.h"
#include "shared_ptr.h"   // the boost::shared_ptr class



/** Macro for constructing a CPPDOM error */
//std::string(__FILE__) + std::string(__LINE__)
#define CPPDOM_ERROR(error_code, local_desc) cppdom::Error(error_code, local_desc, __FILE__, __LINE__ )

//! namespace of the cppdom project
namespace cppdom
{
   // Helper methods
   template <class T> inline void ignore_unused_variable_warning(const T&) { }

   // True if there are characters references: ex: &amp;
   bool textContainsXmlEscaping(const std::string& data);

   // True if there are chars needing escaping
   bool textNeedsXmlEscaping(const std::string& data, bool isCdata);

    // Remove escaping from xml text
   std::string removeXmlEscaping(const std::string& data, bool isCdata);

   // Add escaping to xml text
   std::string addXmlEscaping(const std::string& data, bool isCdata);

   //! xml parsing error codes enumeration
   enum ErrorCode
   {
      xml_unknown = 0,              /**< unspecified or unknown error */
      xml_instream_error,           /**< error in the infile stream */
      xml_opentag_expected,         /**< expected an open tag literal '<' */
      xml_opentag_cdata_expected,   /**< expected a '<' or cdata */
      xml_closetag_expected,        /**< expected a '>' closing tag literal */
      xml_pi_doctype_expected,      /**< expected a processing instruction or doctype tag */
      xml_tagname_expected,         /**< expected a tag name after '<' or '</' */
      xml_closetag_slash_expected,  /**< expected a '/' after closing tag literal '<' */
      xml_tagname_close_mismatch,   /**< tag name from start and end tag mismatch */
      xml_attr_equal_expected,      /**< expected '=' after attribute name */
      xml_attr_value_expected,      /**< expected value after an '=' in attribute */
      xml_save_invalid_nodetype,    /**< invalid nodetype encountered while saving */

      xml_invalid_operation,        /**< Attempted to execute an xml operation that would cause invalid structure */
      xml_invalid_argument,         /**< Attempted to pass an invalid argument */
   // added by kevin for 0.7 compatibility...
      xml_filename_invalid,
      xml_file_access,
      xml_escaping_failure,         /**< Problem with escaping */

      xml_dummy                     /**< dummy error code */
   };
   
   // typedefs
   /** smart pointer to node */
   class Node;
   typedef cppdom_boost::shared_ptr<cppdom::Node> NodePtr;
   
   class Document;
   typedef cppdom_boost::shared_ptr<cppdom::Document> DocumentPtr;


   /** list of node smart pointer */
   typedef std::vector<NodePtr> NodeList;
   typedef NodeList::iterator NodeListIterator;


   // classes

   /**
    * xml error class
    * contains an ErrorCode and is thrown while parsing xml input
    */
   class CPPDOM_CLASS Error
   {
   public:
      /** constructor */
      Error(ErrorCode code, std::string localDesc, std::string location);

      /** constructor that can take line num */
      Error(ErrorCode code, std::string localDesc, std::string file, unsigned line_num);

      /** returns the error code */
      ErrorCode getError() const;

      /** returns the string representation of the error code */
      std::string getStrError() const;

      std::string getString() const;

      /** return additional error info */
      std::string getInfo() const;

   protected:
      ErrorCode   mErrorCode;
      std::string mLocalDesc; /**< Local description of the error */
      std::string mLocation;  /**< The location text */
   };


   /**
    * xml stream position
    * represents the position in the xml input stream; usable if load()
    *    throws an error on parsing xml content
    */
   class CPPDOM_CLASS Location
   {
   public:
      /** Constructor */
      Location();

      /** returns current line */
      int getLine() const;

      /** returns current position in a line */
      int getPos() const;

      /** advances a char */
      void step(int chars = 1);

      /** indicates entering a new line */
      void newline();

      /** reset location */
      void reset();

   protected:
      int mLine;
      int mPos;
   };


   // typedefs
   /** smart pointer for Context */
   typedef cppdom_boost::shared_ptr<class Context> ContextPtr;

   /** smart pointer to the event handler */
   typedef cppdom_boost::shared_ptr<class EventHandler> EventHandlerPtr;

   /**
    * xml parsing context class.
    * the class is the parsing context for the parsed xml document.
    * the class has a tagname lookup table and an entity map
    */
   class CPPDOM_CLASS Context
   {
   public:
      /** ctor */
      Context();
      /** dtor */
      virtual ~Context();

      /** returns the tagname by the tagname handle */
      std::string getTagname(TagNameHandle handle);

      /** inserts a tag name and returns a tag name handle to the string */
      TagNameHandle insertTagname(const std::string& tagname);

      /** returns the current location in the xml stream */
      Location& getLocation();

      /** @name event handling methods */
      //@{
      /** sets the event handler; enables handling events */
      void setEventHandler(EventHandlerPtr ehptr);

      /** returns the currently used eventhandler (per reference) */
      EventHandler& getEventHandler();

      /** returns if a valid event handler is set */
      bool hasEventHandler() const;
      //@}

   protected:
      bool              mInit;            /**< indicates if init_context() was already called */
      int               mNextHandle;      /**< next available tagname handle */
      TagNameMap_t      mTagToName;       /**< matches TagNameHandles to the real std::string's */
      NameToTagMap_t    mNameToTag;       /**< Map name to tag for inserting. */
      Location          mLocation;        /**< location of the xml input stream */
      bool              mHandleEvents;    /**< indicates if the event handler is used */
      EventHandlerPtr   mEventHandler;    /**< current parsing event handler */
   };
   

   /**
    * XML attribute class.
    * Just wraps a string (this is really just and attribute VALUE)
    */
   class CPPDOM_CLASS Attribute
   {
   public:
      Attribute();
      Attribute(const Attribute& attr);
      Attribute(const std::string& val);

#ifndef CPPDOM_NO_MEMBER_TEMPLATES
      template<class T>
      Attribute(const T& val)
      {
         setValue<T>(val);
      }
#endif // ! CPPDOM_NO_MEMBER_TEMPLATES

      const std::string& getString() const;

#ifndef CPPDOM_NO_MEMBER_TEMPLATES
      /**
       * Set mData to the string value of val
       * @note Requires a stream operation of type T
       */
      template<class T>
      void setValue(const T& val)
      {
         std::ostringstream oss;
         oss << val;
         mData = oss.str();
      }

      template<class T>
      T getValue() const
      {
         T t;
         std::istringstream iss(mData);
         iss >> t;
         return t;
      }

      // Specializations of getValue<T> placed inline for Visual Studio 7.
      // MIPSpro and GCC do not handle this. They get out-of-line
      // specializations, found below.
#ifdef _MSC_VER
      /**
       * Specialization of getValue<T> for std::string so that more than just the
       * first word of the attribute string is returned.
       */
      template<>
      std::string getValue<std::string>() const
      {
         return mData;
      }
#endif // ! _MSC_VER
#endif // ! CPPDOM_NO_MEMBER_TEMPLATES

      /** Autoconversion to string (so old code should work) */
      operator std::string() const;

   protected:
      std::string mData;
   };

#ifndef CPPDOM_NO_MEMBER_TEMPLATES
#ifndef _MSC_VER
   template<>
   inline std::string Attribute::getValue<std::string>() const
   {
      return mData;
   }
#endif // ! _MSC_VER
#endif // ! CPPDOM_NO_MEMBER_TEMPLATES


   /**
    * XML tag attribute map.
    * Contains all attributes and values a tag has, represented in a map.
    * maps: attrib_name:string --> attrib_value:string
    */
   class CPPDOM_CLASS Attributes
   {
      friend class Parser;
   public:
      typedef std::map<std::string, std::string> attr_map_t;
      typedef attr_map_t::const_iterator  const_iterator;
      typedef attr_map_t::iterator        iterator;
      typedef attr_map_t::value_type      value_type;

      /** ctor */
      Attributes();

      /**
       * Get the named attribute.
       * @returns empty string "" if not found, else the value.
       */
      std::string get(const std::string& key) const;

      /**
       * Sets new attribute value.
       * If not found, then just insert the new attribute.
       */
      void set(const std::string& key, const std::string& value);

      /**
       * Check if the attribute map has the given attribute.
       * @return false if not found.
       */
      bool has(const std::string& key) const;

      std::pair<iterator, bool> insert(const value_type& x)
      {
         return mMap.insert(x);
      }

      iterator begin()
      {
         return mMap.begin();
      }

      iterator end()
      {
         return mMap.end();
      }

      const_iterator begin() const
      {
         return mMap.begin();
      }

      const_iterator end() const
      {
         return mMap.end();
      }

      attr_map_t::size_type size() const
      {
         return mMap.size();
      }

   private:
      attr_map_t mMap;
   };


   /** xml node.
   * A node has the following properties
   * name - The element name of the node
   * type - The type of the node. see NodeType
   * children - Child elements of the node
   * cdata - the cdata content if of type cdata
   *
   */
   class CPPDOM_CLASS Node
   {
   public:
         /** node type enumeration */
      enum Type
      {
         xml_nt_node,      /**< normal node, can contain subnodes */
         xml_nt_leaf,      /**< a leaf node, which contains no further nodes, eg. <img/> */
         xml_nt_document,  /**< document root node */
         xml_nt_cdata      /**< cdata node, which only contains char data */
      };

      friend class Parser;
   protected:
      /** Default Constructor */
      Node();

   public:
      /** constructor, takes xml context pointer */
      explicit Node(ContextPtr pctx);

      /** Construct a node with a given name */
      explicit Node(std::string nodeName, ContextPtr ctx);

      Node(const Node& node);

      ~Node();

   public:
      /** assign operator */
      Node& operator=(const Node& node);

      /** Returns true if the nodes are equal
      * @param ignoreAttribs - Attributes to ignore in the comparison
      * @param ignoreElements - Elements to ignore in the comparison
      */
      bool isEqual(NodePtr otherNode, const std::vector<std::string>& ignoreAttribs,
                   const std::vector<std::string>& ignoreElements,
                   bool dbgit=false, const unsigned debugIndent=0);

      bool isEqual(NodePtr otherNode)
      {
         std::vector<std::string> empty_strings;
         return isEqual(otherNode, empty_strings, empty_strings );
      }

      /** Returns the local name of the node (the element name) */
      std::string getName();
      /** set the node name */
      void setName(const std::string& name);

      /** @name Type information */
      //@{
      /** returns type of node */
      Node::Type getType() const;

      bool isNode()
      { return getType() == xml_nt_node;}
      bool isLeaf()
      { return getType() == xml_nt_leaf;}
      bool isDocument()
      { return getType() == xml_nt_document;}
      bool isCData()
      { return getType() == xml_nt_cdata;}

      /** sets new nodetype */
      void setType(Node::Type type);
      //@}

      /** @name Attribute information */
      //@{
      /** Check if the node has a given attribute */
      bool hasAttribute(const std::string& name) const;

      /**
       * Get the named attribute
       * @returns empty string ("") if not found, else the value
       * @post Object does not change.
       */
      Attribute getAttribute(const std::string& name) const;

      /** returns attribute map of the node */
      Attributes& getAttrMap();

      /** returns attribute map of the node */
      const Attributes& getAttrMap() const;

      /**
       * sets new attribute value
       * @param attr - Attribute name to set.  There must not be ANY spaces in this name
       * @post Element.attr is set to value.  If it didn't exist before, now it does.
       */
      void setAttribute(const std::string& attr, const Attribute& value);
      //@}

      /** @name Children and parents */
      //@{
      /** Returns true if the node has a child of the given name.
       * @param name    Name can be a single element name or a chain of the form "tag/tag/tag"
       */
      bool hasChild(const std::string& name);

      /** Returns the first child of the given local name.
       */
      NodePtr getChild(const std::string& name);

      /** Return first child of the given name.
       * @param name    Name can be a single element name or a chain of the form "tag/tag/tag"
       */
      NodePtr getChildPath(const std::string& path);

      /** returns a list of the nodes children */
      NodeList& getChildren();

      /**
       * Returns a list of all children (one level deep) with local name of childName
       * \note currently no path-like childname can be passed, like in e.g. msxml
       * If has standard compose() functions, could impl this by calling getChildren(pred)
       */
      NodeList getChildren(const std::string& name);

      /**
       * Returns list of all children.
       * @see getChildren
       * @note Needed because of predicate template matching char*
       */
      NodeList getChildren(const char* name);

      /** Return a list of children that pass the given STL predicate */
      template<class Predicate>
      NodeList getChildrenPred(Predicate pred)
      {
         NodeList result(0);
         NodeList::const_iterator iter;

         // search for all occurances of nodename and insert them into the new list
         for(iter = mNodeList.begin(); iter != mNodeList.end(); ++iter)
         {
            if (pred(*iter))
            {
               result.push_back(*iter);
            }
         }
         return result;
      }

      /**
       * Returns our parent.  The returned value could be NULL.
       *
       * @return Our parent, which may be NULL if we have no parent.
       */
      Node* getParent() const;

      void addChild(NodePtr& node);
      bool removeChild(NodePtr& node);
      bool removeChild(std::string& childName);
      bool removeChildren(std::string& childName);

      //@}

      /** @name CData methods */
      //@{
      /**
       * returns cdata string
       * @note: For node type "cdata", this returns the local cdata.
       *        For other nodes, this attempts to find the first child cdata
       *        node and returns its data.
       */
      std::string getCdata();

      /**
       * Returns the full cdata of the node or immediate children.
      * @note: For node type "cdata", this returns the local cdata.
      *        For other nodes, combines the cdata of all cdata children.
      */
      std::string getFullCdata();

      /** Sets the node cdata.
      * @post For cdata type nodes, this sets the contained cdata
      *       For other types, this sets the cdata of the first cdata node.
      *       If none exists, then one is created called "cdata".
      */
      void setCdata(const std::string& cdata);
      //@}


      /** @name load/save functions */
      //@{
      /** loads xml node from input stream */
      void load(std::istream& in, ContextPtr& context);

      /** saves node to xml output stream
      * @param indent - The amount to indent
      * @doIndent - If true, then indent the output
      * @doNewline - If true then use newlines in output
      */
      void save(std::ostream& out, int indent=0, bool doIndent=true, bool doNewline=true);
      //@}

      /** Returns the context used for this node. */
      ContextPtr getContext();

   protected:
      TagNameHandle  mNodeNameHandle;  /**< handle to the real tag name */

//#ifdef CPPDOM_DEBUG
      std::string    mNodeName_debug;  /**< The node name for debugging */
//#endif
      ContextPtr     mContext;         /**< smart pointer to the context class */
      Node::Type     mNodeType;        /**< The type of the node */
      Attributes     mAttributes;      /**< Attributes of the element */
      std::string    mCdata;           /**< Character data (if there is any) */
      NodeList       mNodeList;        /**< stl list with subnodes */
      Node*          mParent;          /**< Our parent */
   };


   /** XML document root node.
    *
    * Structure of a cppdom document
    *
    * Nested tree of nodes.  Each node has a type (Node::Type).
    * Root node is of type document.  Standard nodes are type node.
    * Under standard nodes there can be cdata nodes.  These nodes are nodes
    * for containing the raw text inside an element.
    *
    */
   class CPPDOM_CLASS Document: public Node
   {
      friend class Parser;
   public:
      Document();

      ~Document();

      /** constructor taking xml context pointer */
      explicit Document(ContextPtr context);

      /** constructor taking xml context pointer */
      explicit Document(std::string docName, ContextPtr context);

      /** returns a list of processing instruction nodes */
      NodeList& getPiList();

      /** returns a list of document type definition rules to check the xml file */
      NodeList& getDtdList();

      /** loads xml Document (node) from input stream */
      void load(std::istream& in, ContextPtr& context);

      /** saves node to xml output stream
      * @param doIndent - If true, then indent the output.
      * @param doNewline - If true, then use newlines in the output.
      */
      void save(std::ostream& out, bool doIndent=true, bool doNewline=true);

      /**
       * \exception throws cppdom::Error when the file name is invalid.
       */
      void loadFile(const std::string& filename) throw(Error);
      void loadFileChecked(const std::string& filename);

      /** Save the document to the given filename.
      * @todo Fix this method.  It doesn't work
      */
      void saveFile(std::string filename);


   protected:
      /** node list of parsed processing instructions */
      NodeList mProcInstructions;

      /** node list of document type definition rules */
      NodeList mDtdRules;
   };

   /** Interface for xml parsing event handler */
   class CPPDOM_CLASS EventHandler
   {
   public:
      /** ctor */
      EventHandler() {}

      /** virtual dtor */
      virtual ~EventHandler() {}

      /** called when parsing of an xml document starts */
      virtual void startDocument() {}

      /** called when ended parsing a document */
      virtual void endDocument() {}

      /** called when parsing a processing instruction */
      virtual void processingInstruction(Node& pinode)
      {  cppdom::ignore_unused_variable_warning(pinode);}

      /** called when start parsing a node */
      virtual void startNode(const std::string& nodename)
      { cppdom::ignore_unused_variable_warning(nodename); }
      /** called when an attribute list was parsed */
      virtual void parsedAttributes(Attributes& attr)
      { cppdom::ignore_unused_variable_warning(attr); }
      /** called when parsing of a node was finished */
      virtual void endNode(Node& node)
      { cppdom::ignore_unused_variable_warning(node);}

      /** called when a cdata section ended */
      virtual void gotCdata(const std::string& cdata)
      { cppdom::ignore_unused_variable_warning(cdata); }
   };


   // ----------------------------------- //
   /** @name Helper methods */
   //@{
   /** Merges data from one node to another.
    * @param fromNode   Node to read data from.
    * @param toNode     Node to merge data onto.
    * @post All elements and attributes in fromNode will exist in toNode.
    */
   void merge(NodePtr fromNode, NodePtr toNode);
   //@}
}

#endif
